cpp/cycles_module/algorithm/cycles.cpp

```diff
@@ -0,0 +1,201 @@
+  for (const auto [from, to] : non_st_edges) {
+    fundamental_cycles->emplace_back(FindFundametalCycle(from, to, state));
+  }
+}
+
+std::vector<std::uint64_t> FindFundametalCycle(std::uint64_t node_a, std::uint64_t node_b, const NodeState &state) {
+  std::vector<std::uint64_t> cycle;
+  if (state.depth[node_a] < state.depth[node_b]) {
+    std::swap(node_a, node_b);
+  }
+
+  // climb until a and b reach the same depth:
+  //
+  //       ()               ()
+  //      /  \             /  \.
+  //    ()   (b)   -->   (a)  (b)
+  //   /  \             /   \.
+  // (a)  ()           ()   ()
+  cycle.emplace_back(node_a);
+  while (state.depth[node_a] > state.depth[node_b]) {
+    node_a = state.parent[node_a];
+    cycle.emplace_back(node_a);
+  }
+
+  if (node_a != node_b) {
+    throw std::runtime_error("There should be no cross edges in DFS tree of an undirected graph.");
+  }
+
+  // Close the cycle
+  cycle.emplace_back(cycle[0]);
+  return cycle;
+}
+
+void CombineCycles(int mask, const std::vector<std::vector<std::uint64_t>> &fundamental_cycles,
+                   const mg_graph::GraphView<> &graph, std::vector<std::vector<mg_graph::Node<>>> *cycles) {
+  std::map<std::pair<std::uint64_t, std::uint64_t>, std::uint64_t> edge_cnt;
+  for (std::size_t i = 0; i < fundamental_cycles.size(); ++i) {
+    if ((mask & (1 << i)) == 0) continue;
+    for (std::size_t j = 1; j < fundamental_cycles[i].size(); ++j) {
+      auto edge = std::minmax(fundamental_cycles[i][j], fundamental_cycles[i][j - 1]);
+      edge_cnt[edge]++;
+    }
+  }
+
+  std::unordered_map<std::uint64_t, std::vector<std::uint64_t>> adj_list;
+  std::unordered_set<std::uint64_t> nodes;
+  for (const auto [key, value] : edge_cnt) {
+    if (value % 2 == 0) continue;
+
+    const auto [from, to] = key;
+    adj_list[from].push_back(to);
+    adj_list[to].push_back(from);
+    nodes.insert(from);
+    nodes.insert(to);
+  }
+
+  // deg(v) = 2 for all vertices in a cycle.
+  for (const auto node : nodes) {
+    if (adj_list[node].size() != 2) return;
+  }
+
+  std::unordered_set<std::uint64_t> visited;
+  std::vector<mg_graph::Node<>> cycle;
+  auto curr_node = *nodes.begin();
+
+  while (visited.find(curr_node) == visited.end()) {
+    cycle.push_back({curr_node});
+    visited.insert(curr_node);
+    for (const auto next_node : adj_list[curr_node]) {
+      if (visited.find(next_node) == visited.end()) curr_node = next_node;
+    }
+  }
+
+  for (const auto node : nodes) {
+    if (visited.find(node) == visited.end()) {
+      return;
+    }
+  }
+
+  if (cycle.size() > 2) {
+    cycles->emplace_back(cycle);
+  }
+}
+
+void GetCyclesFromFundamentals(const std::vector<std::vector<uint64_t>> &fundamental_cycles,
+                               const mg_graph::GraphView<> &graph, std::vector<std::vector<mg_graph::Node<>>> *cycles) {
+  auto size = 1 << fundamental_cycles.size();
+  for (std::int32_t mask = 1; mask < size; ++mask) {
```

Comment:
I would use `std::uint32_t` for mask stuff. It's more natural to use unsigned types when doing bit stuff.